packages/api/src/external/commonwell/patient.ts

```diff
@@ -508,51 +478,214 @@ async function validateCWEnabled({
-    const resetPatient = await resetPatientScheduledDocQueryRequestId({
+  if (foundNewDemographicsHere) {
+    await updatePatientLinkDemographics({
+      requestId,
       patient: updatedPatient,
       source: MedicalDataSource.COMMONWELL,
+      links: newDemographicsHere,
     });
+  }
+  await update({
+    patient: updatedPatient,
+    facilityId,
+    getOrgIdExcludeList,
+    rerunPdOnNewDemographics: false,
+  });
+  analytics({
+    distinctId: updatedPatient.cxId,
+    event: EventTypes.rerunOnNewDemographics,
+    properties: {
+      hie: MedicalDataSource.COMMONWELL,
+      patientId: updatedPatient.id,
+      requestId,
+      foundNewDemographicsHere,
+      foundNewDemographicsAcrossHies,
+    },
+  });
+  return true;
+}
+
+export async function runNexPdIfScheduled({
+  patient,
+  requestId,
+}: {
+  patient: Patient;
+  requestId: string;
+}): Promise<boolean> {
+  const updatedPatient = await getPatientOrFail(patient);
+
+  const scheduledPdRequest = getCWData(updatedPatient.data.externalData)?.scheduledPdRequest;
+  if (!scheduledPdRequest) {
+    return false;
+  }
+
+  await resetPatientScheduledPatientDiscoveryRequestId({
+    patient,
+    source: MedicalDataSource.COMMONWELL,
+  });
+  await update({
+    patient: updatedPatient,
+    facilityId: scheduledPdRequest.facilityId,
+    getOrgIdExcludeList: scheduledPdRequest.getOrgIdExcludeList ?? getCqOrgIdsToDenyOnCw,
+    requestId: scheduledPdRequest.requestId,
+    forceCWUpdate: scheduledPdRequest.forceCommonwell,
+    rerunPdOnNewDemographics: scheduledPdRequest.rerunPdOnNewDemographics,
+  });
+  analytics({
+    distinctId: updatedPatient.cxId,
+    event: EventTypes.runScheduledPatientDiscovery,
+    properties: {
+      hie: MedicalDataSource.COMMONWELL,
+      patientId: updatedPatient.id,
+      requestId,
+      scheduledPdRequestId: scheduledPdRequest.requestId,
+    },
+  });
+  return true;
+}
 
+export async function queryDocsIfScheduled({
+  patientIds,
+  getOrgIdExcludeList,
+  isFailed = false,
+}: {
+  patientIds: Pick<Patient, "id" | "cxId">;
+  getOrgIdExcludeList: () => Promise<string[]>;
+  isFailed?: boolean;
+}): Promise<void> {
+  const patient = await getPatientOrFail(patientIds);
+
+  const scheduledDocQueryRequestId = getCWData(
+    patient.data.externalData
+  )?.scheduledDocQueryRequestId;
+  if (!scheduledDocQueryRequestId) {
+    return;
+  }
+
+  await resetPatientScheduledDocQueryRequestId({
+    patient,
+    source: MedicalDataSource.COMMONWELL,
+  });
+  if (isFailed) {
+    await setDocQueryProgress({
+      patient,
+      requestId: scheduledDocQueryRequestId,
+      source: MedicalDataSource.COMMONWELL,
+      downloadProgress: { status: "failed", total: 0 },
+      convertProgress: { status: "failed", total: 0 },
+    });
+  } else {
     await queryAndProcessDocuments({
-      patient: resetPatient,
+      patient,
       requestId: scheduledDocQueryRequestId,
       getOrgIdExcludeList,
     });
```

Comment:
This is a long running function and we should probably call it asynchronously.

Reply:
Done.